echopop/acoustics.py

```diff
@@ -0,0 +1,452 @@
+
+    # Correct the acoustic survey transect intervals
+    nasc_interval_df = correct_transect_intervals(input_dict["acoustics"]["nasc_df"])
+
+    # Select the appropriate NASC column based on the inclusion or exclusion of age-1 fish
+    if settings_dict["transect"]["exclude_age1"]:
+        # ---- Drop the unused column
+        nasc_interval_df.drop("NASC_all_ages", axis=1, inplace=True)
+        # ---- Rename the used column
+        nasc_interval_df.rename(columns={"NASC_no_age1": "nasc"}, inplace=True)
+        # ---- Calculate age-1 NASC and weight proportions
+        age1_proportions = age1_metric_proportions(
+            input_dict["biology"]["distributions"],
+            analysis_dict["biology"]["proportions"],
+            configuration_dict["TS_length_regression_parameters"]["pacific_hake"],
+            settings_dict,
+        )
+        # ---- Calculate adult proportions
+        # -------- Initialize dataframe
+        adult_proportions = age1_proportions.copy()
+        # -------- Invert proportions
+        adult_proportions.iloc[:, np.arange(1, 4)] = (
+            1.0 - adult_proportions.iloc[:, np.arange(1, 4)]
+        )
+
+    else:
+        # ---- Drop the unused column
+        nasc_interval_df.drop("NASC_no_age1", axis=1, inplace=True)
+        # ---- Rename the used column
+        nasc_interval_df.rename(columns={"NASC_all_ages": "nasc"}, inplace=True)
+        # ---- Assign filled adult proportions dataframe
+        adult_proportions = pd.DataFrame(
+            {
+                f"{stratum_col}": np.unique(nasc_interval_df[stratum_col]),
+                "number_proportion": np.ones(len(np.unique(nasc_interval_df[stratum_col]))),
+                "weight_proportion": np.ones(len(np.unique(nasc_interval_df[stratum_col]))),
+                "nasc_proportion": np.ones(len(np.unique(nasc_interval_df[stratum_col]))),
+            }
+        )
+
+    # Merge hake fraction data into `nasc_interval_df`
+    # ---- Initial merge
+    nasc_interval_df = nasc_interval_df.merge(
+        input_dict["spatial"]["strata_df"], on=[stratum_col, "haul_num"], how="outer"
+    )
+    # ---- Replace `fraction_hake` where NaN occurs
+    nasc_interval_df["fraction_hake"] = nasc_interval_df["fraction_hake"].fillna(0.0)
+    # ---- Drop NaN
+    nasc_interval_df.dropna(subset=["transect_num"], inplace=True)
+
+    # Calculate the along-transect number density (animals per nmi^2)
+    # ---- Merge NASC measurements with mean sigma_bs for each stratum
+    nasc_biology = nasc_interval_df.merge(sigma_bs_strata, on=[stratum_col])
+    # ---- Calculate the number densities
+    nasc_biology["number_density"] = (
+        nasc_biology["fraction_hake"]
+        * nasc_biology["nasc"]
+        / (4.0 * np.pi * nasc_biology["sigma_bs_mean"])
+    )
+    # ---- Round to the nearest whole number (we don't want fractions of a fish)
+    nasc_biology["number_density"] = np.round(nasc_biology["number_density"])
```

Comment:
I think fraction is fine when we talk about density? Even if you force number density to be integer per nmi^2, it will still be non-integer per km^2, for example.

Reply:
Here I was just trying to replicate the `EchoPro` code that explicitly rounds the acoustically estimated $\rho_{N}$ values. Otherwise, I'd agree with your statement. In general, I think that the differences produced by the choice to round to the closest whole individual (per unit area) versus not are extremely small. Moreover, whatever amount of uncertainty is introduced with either treatment is going to be massively outsized by the uncertainty introduced from the calculation of mean $\sigma_{bs}$ for each stratum.

Reply:
yeah, I agree the difference would be minuscule. But since this is the case, how about just removing that `np.round`? It is more straightforward (and correct IMHO) without it.